File: codigo/p2.c
/*
Programa 2

Multiiplicar duas matrizes quadráticas de dimensões NxN.

*/

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {

	int n;
	if (argc == 2) {
		n = atoi(argv[1]);
	}

	srand(time(NULL));

	// Para vetores grandes é necessário utilizar alocação dinâmica
	// Necessário desalocar a memória antes do final do programa.
	double** m1 = (double**)malloc(n * sizeof(double));
	double** m2 = (double**)malloc(n * sizeof(double));
	double** mm = (double**)malloc(n * sizeof(double));

	int i,j;
	for (i = 0; i < n; i++) {
		m1[i] = (double*)malloc(n * sizeof(double));
		m2[i] = (double*)malloc(n * sizeof(double));
		mm[i] = (double*)malloc(n * sizeof(double));

		for (j = 0; j < n; j++) {
			m1[i][j] = ((double)rand()/(double)RAND_MAX);
			m2[i][j] = ((double)rand()/(double)RAND_MAX);
		}
	}


	void multiMatriz(double** mm, double** m1, double** m2, int l1, int c1, int l2, int c2) {
		int i, j, a;

		// if c1 == l2

		for (i = 0; i < l1; i++) {
			for (j = 0; j < c2; j++) {
        
				mm[i][j] = 0;

				for (a = 0; a < c1; a++) {
					mm[i][j] += (m1[i][a] * m2[a][j]);
				}
			}
		}
		
	}


	void printMatriz(double** mm, int l, int c) {
		int i,j;
		

		for (i = 0; i < l; i++) {
			for (j = 0; j < c; j++) {
				printf("%f\t", mm[i][j]);
			}
			printf("\n");
		}
	}

 	multiMatriz(mm, m1, m2, n, n, n, n);
	printMatriz(mm, n, n);

	for (i = 0; i < n; i++) {
		free(m1[i]);
		free(m2[i]);
		free(mm[i]);
	}
	free(m1); free(m2); free(mm);

	return 0;
}
	

File: codigo/p2.py
# Programa 2

# Multiiplicar duas matrizes quadráticas de dimensões NxN.

import sys
import random


def multiplica_matriz(m1, m2):
  mr = []
  for i in range(len(m1)):
    mr.append([])
    for j in range(len(m2[0])):
      mr[i].append(sum([x*y for x,y in zip(m1[i],m2[j])]))

  return mr


def main(argv):
	n = int(argv)
	matriz1 = [[random.random() for _ in range(n)] for _ in range(n)]
	matriz2 = [[random.random() for _ in range(n)] for _ in range(n)]

	print(multiplica_matriz(matriz1, matriz2))


if __name__ == "__main__":
	if len(sys.argv) != 2:
		print("Erro")
		sys.exit()

	main(sys.argv[1])